NeoML/src/Dnn/DnnLora.cpp

```diff
@@ -0,0 +1,327 @@
+}
+
+//----------------------------------------------------------------------------------------------------------------------
+
+CLoraBuilder::CLoraBuilder()
+{
+	// Default layers supported by LoRA
+	compositeClasses.Add( { "NeoMLDnnTransformerEncoderLayer", "NeoMLDnnMultiheadAttentionLayer",
+		"FmlCnnCompositeLayer", "FmlCnnRecurrentLayer", "FmlCnnTemplateLayer", "NeoMLTemplateLayerExt" } );
+}
+
+CLoraBuilder::CLoraBuilder( const CArray<CString>& _compositeClases )
+{
+	_compositeClases.CopyTo( compositeClasses );
+}
+
+void CLoraBuilder::BuildFcWrapper( CDnnLayerGraph& graph, const char* fcName, const CLoraParams& params ) const
+{
+	NeoAssert( graph.HasLayer( fcName ) );
+	CPtr<CFullyConnectedLayer> fc = CheckCast<CFullyConnectedLayer>( graph.GetLayer( fcName ) );
+	graph.DeleteLayer( *fc );
+
+	NeoAssert( fc->Weights() != nullptr ); // "LoRA for uninitialized fully-connected" doesn't make sense 
+	CPtr<CLoraFullyConnectedLayer> loraFc = FINE_DEBUG_NEW CLoraFullyConnectedLayer( *fc->Weights(),
+		fc->FreeTerms(), params );
+	loraFc->SetName( fc->GetName() );
+	graph.AddLayer( *loraFc );
+	loraFc->Connect( 0, fc->GetInputName( 0 ), fc->GetInputOutputNumber( 0 ) );
+}
+
+int CLoraBuilder::BuildAllFcWrappers( CDnnLayerGraph& rootGraph, const CLoraParams& params ) const
+{
+	auto impl = [this, &params] ( CDnnLayerGraph& currGraph, auto&& impl ) -> int
+	{
+		int result = 0;
+
+		CArray<const char*> layerNames;
+		currGraph.GetLayerList( layerNames );
+		for( const char* layerName : layerNames ) {
+			CPtr<CBaseLayer> layer = currGraph.GetLayer( layerName );
+
+			CFullyConnectedLayer* fc = dynamic_cast<CFullyConnectedLayer*>( layer.Ptr() );
+			if( fc != nullptr ) {
+				BuildFcWrapper( currGraph, layerName, params );
+				++result;
+				continue;
+			}
+
+			const CString layerClass = GetLayerClass( *layer );
+			if( compositeClasses.Find( layerClass ) != NotFound ) {
+				CCompositeLayer* composite = dynamic_cast<CCompositeLayer*>( layer.Ptr() );
+				NeoAssert( composite != nullptr );
+				result += impl( *composite, impl ); // recurrent call of this lambda
+			}
+		}
+
+		return result;
+	};
+
+	return impl( rootGraph, impl );
+}
+
+int CLoraBuilder::DisableNonLoraTraining( CDnnLayerGraph& graph ) const
+{
+	auto impl = [] ( CDnnLayerGraph& currGraph, auto&& impl ) -> int
+	{
+		int result = 0;
+
+		CArray<const char*> layerNames;
+		currGraph.GetLayerList( layerNames );
+		for( const char* layerName : layerNames ) {
+			CPtr<CBaseLayer> layer = currGraph.GetLayer( layerName );
+
+			CLoraFullyConnectedLayer* loraFc = dynamic_cast<CLoraFullyConnectedLayer*>( layer.Ptr() );
+			if( loraFc != nullptr ) {
+				continue; // don't touch LoRA wrappers
+			}
+
+			CCompositeLayer* composite = dynamic_cast<CCompositeLayer*>( layer.Ptr() );
+			if( composite != nullptr ) {
+				result += impl( *composite, impl );
+				continue;
+			}
+
+			if( layer->IsLearnable() && layer->IsLearningEnabled() ) {
+				layer->DisableLearning();
```

Comment:
Do we really need to disable learning of all layers in dnn, not only baseFc inside LoraFullConnectedLayer? This may problem of training lookUpLayer (embeddings)..

Reply:
I think so. The whole pipeline suggests that we can store only LoRA weights and easily merging them into large pre-trained network. And the downside of your suggestion is that you will have to keep a copy of embeddings into every LoRA archive. Plus embeddings themselves could be quite large in contrast with the idea of LoRA to _reduce memory consumption by freezing large matrices of pre-trained network and training only small additional matrices_.